Add Configurable RTC options

RisingFog opened the pull request:
![](https://i.gyazo.com/ada31f69d6073e779fb17610604d8a61.png)

This allows a way to customize what your emulation time when starting a game. This is most useful for TASes which rely on the RTC being set to a particular seed for RNG. Will fix issue 9693: https://bugs.dolphin-emu.org/issues/9693

The majority of the functionality is here, but there are some TODOs to work on:

- [x] Have the calendar show correctly in GC BIOS
- [x] Limit RTC to valid dates
- [x] Increment clock by realtime instead of emulated time
- [x] Test Movie features
- [x] Add description to GUI
- [x] Decide where to actually place this UI function
- [x] Clean up messy code

<!-- Reviewable:start -->
---
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/dolphin-emu/dolphin/4007)
<!-- Reviewable:end -->

lioncash commented:
Review status: 0 of 8 files reviewed at latest revision, 11 unresolved discussions.

---

*[Source/Core/Common/Timer.cpp, line 199 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbQ9d1NFtDaEYTMPf5:-KMbQ9d2lDiGWEcSwjFJ:132184583) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/Common/Timer.cpp#L199)):*
> ```cpp
> }
> 
> u64 Timer::GetLocalTimeSinceJan1970(u32 d, u32 t)
> ```

These should be `date` and `time`

You may want to consider implementing the no-arg version of this function in terms of the this one. This duplicates quite a bit of code.

---

*[Source/Core/Common/Timer.cpp, line 202 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbQskXt6ZHgcs7wOHX:-KMbQskYt5BJjOnkY8FJ:662580446) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/Common/Timer.cpp#L202)):*
> ```cpp
> {
>   time_t sysTime, customTime, tzDiff, tzDST;
>   struct tm* gmTime;
> ```

`struct` isn't necessary.

---

*[Source/Core/Common/Timer.cpp, line 203 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbQTEoK4C1Ei2BuQFV:-KMbQTEoK4C1Ei2BuQFW:117186160) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/Common/Timer.cpp#L203)):*
> ```cpp
>   time_t sysTime, customTime, tzDiff, tzDST;
>   struct tm* gmTime;
>   customTime = (d + t);
> ```

the declaration can be joined with the assignment here.

---

*[Source/Core/Common/Timer.cpp, line 207 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbQXmpLTiLpL8FZSzW:-KMbQXmpLTiLpL8FZSzX:-469765787) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/Common/Timer.cpp#L207)):*
> ```cpp
> 
>   // Account for DST where needed
>   gmTime = localtime(&sysTime);
> ```

ditto

---

*[Source/Core/Common/Timer.cpp, line 217 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbQwWJhidFwHpBP_qt:-KMbQwWJhidFwHpBP_qu:1869053803) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/Common/Timer.cpp#L217)):*
> ```cpp
>   tzDiff = sysTime - mktime(gmTime);
> 
>   return (u64)(customTime + tzDiff + tzDST);
> ```

`static_cast`

---

*[Source/Core/Common/Timer.h, line 30 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbR0bhZaD6_W_KtJSn:-KMbR0biN4nKcLjEDSf4:1475465092) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/Common/Timer.h#L30)):*
> ```objc
>   static u64 GetTimeSinceJan1970();
>   static u64 GetLocalTimeSinceJan1970();
>   static u64 GetLocalTimeSinceJan1970(u32 date, u32 time);
> ```

There should be documentation explaining how date and time data should be representedin order for this function to work correctly (typically you'd use `<chrono>` to represent time points in C++11 capable codebases as well).

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 110 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbRQJN42kJ9FUp2ryq:-KMbRQJN42kJ9FUp2ryr:-305512373) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L110)):*
> ```cpp
>   m_clock_override_slider->Enable(oc_enabled);
>   UpdateCPUClock();
>   wxDateTime custom_rtc_date = wxDateTime((time_t)SConfig::GetInstance().m_customRTCDate);
> ```

static_cast

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 114 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbRSX4zc-qk4Lw-EDi:-KMbRSX4zc-qk4Lw-EDj:-305512373) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L114)):*
> ```cpp
>   bool custom_rtc_enabled = SConfig::GetInstance().bCustomRTC;
>   m_custom_rtc_checkbox->SetValue(custom_rtc_enabled);
>   if (custom_rtc_date > (time_t)0)
> ```

static_cast

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 157 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbRqE8hRvVvUkIiEoo:-KMbRqE9niWmfvIX05yV:-1404009933) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L157)):*
> ```cpp
>   int* pMin = &min;
>   int* pSec = &sec;
>   m_custom_rtc_time_picker->GetTime(pHour, pMin, pSec);
> ```

You can just take references to the variables directly instead of using pointers and re-dereferencing them back on the original variables.

```cpp
int hour;
int min;
int sec;
m_custom_rtc_time_picker->GetTime(&hour, &min, &sec);
```

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 161 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbS-V6foQlZ0DMj623:-KMbS-V6foQlZ0DMj624:-305512373) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L161)):*
> ```cpp
>   min = *pMin;
>   sec = *pSec;
>   u32 epoch = (u32)((hour * 3600) + (min * 60) + sec);
> ```

static_cast

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.h, line 7 \[r1\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMbSaRW7NB0ONz8xUfR:-KMbSaRW7NB0ONz8xUfS:-1249356326) ([raw file](https://github.com/dolphin-emu/dolphin/blob/ee77de052469657f2c6b3229d5b9f0675e34c2e8/Source/Core/DolphinWX/Config/AdvancedConfigPane.h#L7)):*
> ```objc
> #pragma once
> 
> #include <wx/datectrl.h>
> ```

`wxDatePickerCtrl`, and `wxTimePickerCtrl` can be forward declarations.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 6 files reviewed at latest revision, 12 unresolved discussions, some commit checks failed.

---

*[Source/Core/Core/ConfigManager.h, line 140 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMkJ5wDm6FXeW09bD7B:-KMkJ5wEPL-kW2pWiFhH:851574050) ([raw file](https://github.com/dolphin-emu/dolphin/blob/065a66648404a158379a4be45cd78a8ea5e27ca0/Source/Core/Core/ConfigManager.h#L140)):*
> ```objc
>   // Custom RTC
>   bool bCustomRTC;
>   u32 m_customRTC;
> ```

These names are confusingly similar. Mixing variable naming styles isn't especially nice either.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 8 files reviewed at latest revision, 12 unresolved discussions.

---

*[Source/Core/Core/ConfigManager.h, line 140 \[r2\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMkJ5wDm6FXeW09bD7B:-KMlluewDR5q8AjKVo1U:-872880918) ([raw file](https://github.com/dolphin-emu/dolphin/blob/065a66648404a158379a4be45cd78a8ea5e27ca0/Source/Core/Core/ConfigManager.h#L140)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> These names are confusingly similar. Mixing variable naming styles isn't especially nice either.
</details>
I've renamed them.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

lioncash commented:
Review status: 0 of 10 files reviewed at latest revision, 5 unresolved discussions.

---

*[Source/Core/Common/Timer.h, line 30 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp5ghypnZhkCjhKPHB:-KMp5ghypnZhkCjhKPHC:-716297005) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.h#L30)):*
> ```objc
>   static u64 GetTimeSinceJan1970();
>   static u64 GetLocalTimeSinceJan1970();
>   static u64 GetLocalTimeSinceJan1970(u32 epoch_time);
> ```

What's the representation of epoch_time supposed to be? This should be documented above that function, considering it's in the generic common code of the codebase. If you use std::chrono, you can also enforce that desired representation while also making it self-documenting (e.g. `std::chrono::seconds`, `std::chrono::milliseconds`, etc)

---

*[Source/Core/Core/Movie.cpp, line 520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp64f0laQcVsS1vFN2:-KMp64f0laQcVsS1vFN3:631696262) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/Movie.cpp#L520)):*
> ```cpp
>     s_recordingStartTime = CEXIIPL::NetPlay_GetGCTime();
>   }
>   else if (SConfig::GetInstance().bEnableCustomRTC && SConfig::GetInstance().m_customRTCValue > 0)
> ```

`SConfig::GetInstance().bEnableCustomRTC && SConfig::GetInstance().m_customRTCValue > 0`

should likely be a helper instance boolean function that SConfig exposes. This exact line is used in more than one place.

---

*[Source/Core/Core/HW/SystemTimers.cpp, line 87 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp7aIm33RDxVLsZBqZ:-KMp7aIn3bFqdGArNQE9:-233674129) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/HW/SystemTimers.cpp#L87)):*
> ```cpp
> 
> // Custom RTC
> u64 localtime_on_boot;
> ```

This should be a static variable (and have the `s_` prefix).

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 122 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp57ajkARgAMtOXpRS:-KMp57akcshKax1XuckY:1129333086) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L122)):*
> ```cpp
>   UpdateCPUClock();
> 
>   wxDateTime custom_rtc = wxDateTime(static_cast<time_t>(SConfig::GetInstance().m_customRTCValue));
> ```

This logic should likely go in its own function, `LoadDateTimeValues()` or something, which is then called through LoadGUIValues().

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/Common/Timer.cpp, line 180 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMpCEpbnHIlG_GuviIs:-KMpCEpbnHIlG_GuviIt:-1334943946) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.cpp#L180)):*
> ```cpp
> namespace
> {
> u64 LocalTimeDST(time_t time)
> ```

Time zones aren't mentioned but DST is, even though both are equally important.

---

*[Source/Core/Common/Timer.cpp, line 209 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMpCZ41ovv8lVxDlY0y:-KMpCZ41ovv8lVxDlY0z:-1129555102) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.cpp#L209)):*
> ```cpp
> u64 Timer::GetLocalTimeSinceJan1970(u32 epoch_time)
> {
>   return LocalTimeDST(epoch_time);
> ```

Using any data from the local machine (in this case, DST and time zone) is fishy. Are you sure you're not compensating for something that should have been done in DolphinWX when setting the time?

---

*[Source/Core/Common/Timer.h, line 30 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp5ghypnZhkCjhKPHB:-KMpC5LVJ8YBTsSXA-IQ:68171157) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.h#L30)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> What's the representation of epoch_time supposed to be? This should be documented above that function, considering it's in the generic common code of the codebase. If you use std::chrono, you can also enforce that desired representation while also making it self-documenting (e.g. `std::chrono::seconds`, `std::chrono::milliseconds`, etc)
</details>
This is not an appropriate function name, because the function doesn't return anything related to the local time.

---

*[Source/Core/Core/Movie.cpp, line 520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp64f0laQcVsS1vFN2:-KMpBZRw8ApxlWlekxMF:-1965051935) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/Movie.cpp#L520)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> `SConfig::GetInstance().bEnableCustomRTC && SConfig::GetInstance().m_customRTCValue > 0`
> 
> should likely be a helper instance boolean function that SConfig exposes. This exact line is used in more than one place.
> 
</details>
What is `SConfig::GetInstance().m_customRTCValue > 0` supposed to be doing anyway? What makes 0 invalid as opposed to, say, 1?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
*[Source/Core/Common/Timer.cpp, line 180 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMpCEpbnHIlG_GuviIs:-KMpGgIsfXVDBon7mvgB:1755713581) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.cpp#L180)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Time zones aren't mentioned but DST is, even though both are equally important.
</details>
Renamed.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/Common/Timer.cpp, line 209 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMpCZ41ovv8lVxDlY0y:-KMpGmQQEV3Rdj9KW0b8:-780303486) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.cpp#L209)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Using any data from the local machine (in this case, DST and time zone) is fishy. Are you sure you're not compensating for something that should have been done in DolphinWX when setting the time?
</details>
The code DolphinWX sets the epoch time as current local time of the user. Like regular localtime, it needs to be converted to GMT so that it displays properly on the GC and Wii.

---

*[Source/Core/Common/Timer.h, line 30 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp5ghypnZhkCjhKPHB:-KMpJKRP5wkTrEcLRGZs:-1807174361) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.h#L30)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> This is not an appropriate function name, because the function doesn't return anything related to the local time.
</details>
I've renamed the function to be more appropriate.

---

*[Source/Core/Core/Movie.cpp, line 520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp64f0laQcVsS1vFN2:-KMpJNoj3UZ5kQM1G3q6:1984698687) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/Movie.cpp#L520)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> What is `SConfig::GetInstance().m_customRTCValue > 0` supposed to be doing anyway? What makes 0 invalid as opposed to, say, 1?
</details>
It's supposed to check if customRTCValue is not set but it's enabled. However, it blocks 

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
*[Source/Core/Core/HW/SystemTimers.cpp, line 87 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp7aIm33RDxVLsZBqZ:-KMpKzvUxCpwm5mPhyfq:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/HW/SystemTimers.cpp#L87)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should be a static variable (and have the `s_` prefix).
</details>
Done.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

JosJuice commented:
Review status: 0 of 10 files reviewed at latest revision, 7 unresolved discussions.

---

*[Source/Core/Common/Timer.cpp, line 209 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMpCZ41ovv8lVxDlY0y:-KMpMR_gdN6ZQWcITXpR:-555486324) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.cpp#L209)):*
<details><summary><i>Previously, RisingFog (Chris Burgener) wrote…</i></summary>
> The code DolphinWX sets the epoch time as current local time of the user. Like regular localtime, it needs to be converted to GMT so that it displays properly on the GC and Wii.
</details>
If it's true that it needs to be in GMT for the GC/Wii to display it correctly, the GC/Wii must know the user's time zone so that it can convert GMT to local time. How can it know that?

---

*[Source/Core/Common/Timer.h, line 30 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp5ghypnZhkCjhKPHB:-KMpMD6pZJvwbOrfOVoT:1590931828) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.h#L30)):*
<details><summary><i>Previously, RisingFog (Chris Burgener) wrote…</i></summary>
> I've renamed the function to be more appropriate.
</details>
Your comment in the code says that it converts to localtime, but a comment you wrote on Reviewable says it converts localtime to GMT. Which one is it?

---

*[Source/Core/Core/Movie.cpp, line 520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp64f0laQcVsS1vFN2:-KMpLPuPYAHF3aVyiZPN:1766480887) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/Movie.cpp#L520)):*
<details><summary><i>Previously, RisingFog (Chris Burgener) wrote…</i></summary>
> It's supposed to check if customRTCValue is not set but it's enabled. However, it blocks 
</details>
That sounds weird. It's simpler both for users and the code if there just are two states: Either it's disabled and the time doesn't apply, or it's enabled and the time applies.

Nobody enables the option and then expects it to not do anything because of them not setting the date value. As long as the date value is next to the enable checkbox, the user will see what custom time Dolphin will use, and they will change it or disable the option if they don't want the default one. What would enabled but not set even look like in the GUI?

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 9 files reviewed at latest revision, 7 unresolved discussions, some commit checks failed.

---

*[Source/Core/Common/Timer.cpp, line 209 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMpCZ41ovv8lVxDlY0y:-KMprtdL4W-KNYSELdzD:289194068) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.cpp#L209)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> If it's true that it needs to be in GMT for the GC/Wii to display it correctly, the GC/Wii must know the user's time zone so that it can convert GMT to local time. How can it know that?
</details>
Removed and converting time within DolphinWX.

---

*[Source/Core/Core/Movie.cpp, line 520 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp64f0laQcVsS1vFN2:-KMprp4J5hm1w7lieWuO:-1014529885) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Core/Movie.cpp#L520)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> That sounds weird. It's simpler both for users and the code if there just are two states: Either it's disabled and the time doesn't apply, or it's enabled and the time applies.
> 
> Nobody enables the option and then expects it to not do anything because of them not setting the date value. As long as the date value is next to the enable checkbox, the user will see what custom time Dolphin will use, and they will change it or disable the option if they don't want the default one. What would enabled but not set even look like in the GUI?
</details>
Made it only check if EnableCustomRTC is true.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 122 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp57ajkARgAMtOXpRS:-KMprh12XtTsgiEXUtQg:1964245312) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L122)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This logic should likely go in its own function, `LoadDateTimeValues()` or something, which is then called through LoadGUIValues().
</details>
Moved.

---

*[Source/Core/Common/Timer.h, line 30 \[r4\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMp5ghypnZhkCjhKPHB:-KMprhyinSgfYAGV1qS6:450518856) ([raw file](https://github.com/dolphin-emu/dolphin/blob/af73cfc7f663bfbc265df86aafa77edab3432eda/Source/Core/Common/Timer.h#L30)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> Your comment in the code says that it converts to localtime, but a comment you wrote on Reviewable says it converts localtime to GMT. Which one is it?
</details>
Removed the function entirely, moving it to within DolphinWX.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
@dolphin-emu-bot rebuild

RisingFog commented:
I screwed something up with my logic move to DolphinWX, this is not ready to merge yet.

JosJuice commented:
Review status: 0 of 9 files reviewed at latest revision, 6 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.h, line 38 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyP7kwM4YDWF5hHr2V:-KMyP7kwM4YDWF5hHr2W:968535705) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.h#L38)):*
> ```objc
>   static u32 ConvertToFromLocalTimezone(u32 input_date, bool to_gmt);
>   u32 temp_date;
>   u32 temp_time;
> ```

These two need an `m_` prefix.

Actually, why do these exist at all? You can call m_custom_rtc_date_picker->GetValue().GetTicks()` in the places that use `temp_date` right now, and respectively for time. Keeping around extra state just makes the code harder to reason about.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

lioncash commented:
Review status: 0 of 9 files reviewed at latest revision, 8 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 141 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyUA3HVQM5SyVRGjRS:-KMyUA3IvGHRD39uvy_k:-1934948829) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L141)):*
> ```cpp
> void AdvancedConfigPane::OnCustomRTCCheckBoxChanged(wxCommandEvent& event)
> {
>   SConfig::GetInstance().bEnableCustomRTC = m_custom_rtc_checkbox->IsChecked();
> ```

You can just use the stored value directly to make this a little bit less verbose.

```cpp
const bool checked = m_custom_rtc_checkbox->IsChecked();

SConfig::GetInstance().bEnableCustomRTC = checked;
m_custom_rtc_date_picker->Enable(checked);
m_custom_rtc_time_picker->Enable(checked);
```

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 183 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyRQRxbr6_IcxdqOwj:-KMyRQRyacTUkERQZgE0:414134045) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L183)):*
> ```cpp
>   // Limit dates to valid ranges (2000 to 2099 for GC, 2000 to 2035 for Wii)
>   if (SConfig::GetInstance().bWii)
>     m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Month(0), 2000),
> ```

```cpp
m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Jan, 2000),
                                   wxDateTime(31, wxDateTime::Dec, 2035));
```

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 186 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMySGMG-mTjuBV3m9Vu:-KMySGMG-mTjuBV3m9Vv:835615903) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L186)):*
> ```cpp
>                                        wxDateTime(31, wxDateTime::Month(11), 2035));
>   else
>     m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Month(0), 2000),
> ```

```cpp
m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Jan, 2000),
                                   wxDateTime(31, wxDateTime::Dec, 2099));
```

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 195 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyTT4_lg9bnGd6OLQ0:-KMyTT4_lg9bnGd6OLQ1:-1562020893) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L195)):*
> ```cpp
> {
>   SConfig::GetInstance().m_customRTCValue =
>       ConvertToFromLocalTimezone((temp_date + temp_time), true);
> ```

You could store the result of `temp_date + temp_time` into a const time_t variable to get rid of repetition here.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 202 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyTKbtdPUOe5NYT8Q8:-KMyTKbtdPUOe5NYT8Q9:-144270403) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L202)):*
> ```cpp
> u32 AdvancedConfigPane::ConvertToFromLocalTimezone(u32 input_date, bool to_gmt)
> {
>   // wxWidgets's To/FromUTC() functions, ...well, suck. Instead, we roll our own.
> ```

This should explain why they suck.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 204 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyTC_29b3d52kCg0zc:-KMyTC_29b3d52kCg0zd:-1405260296) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L204)):*
> ```cpp
>   // wxWidgets's To/FromUTC() functions, ...well, suck. Instead, we roll our own.
>   time_t tzDiff, tzDST;
>   time_t inputDate = static_cast<time_t>(input_date);
> ```

You should probably not use names that differ this closely to one another. The parameter itself should likely be a time_t.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

RisingFog commented:
Review status: 0 of 10 files reviewed at latest revision, 8 unresolved discussions.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 141 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyUA3HVQM5SyVRGjRS:-KMz0E-4FevNg436L-_x:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L141)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You can just use the stored value directly to make this a little bit less verbose.
> 
> ```cpp
> const bool checked = m_custom_rtc_checkbox->IsChecked();
> 
> SConfig::GetInstance().bEnableCustomRTC = checked;
> m_custom_rtc_date_picker->Enable(checked);
> m_custom_rtc_time_picker->Enable(checked);
> ```
</details>
Done.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 183 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyRQRxbr6_IcxdqOwj:-KMz0GWyL1L9XPOi5gn5:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L183)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Jan, 2000),
>                                    wxDateTime(31, wxDateTime::Dec, 2035));
> ```
</details>
Done.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 186 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMySGMG-mTjuBV3m9Vu:-KMz0Hv9Z2kZaAYRFzyS:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L186)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> ```cpp
> m_custom_rtc_date_picker->SetRange(wxDateTime(1, wxDateTime::Jan, 2000),
>                                    wxDateTime(31, wxDateTime::Dec, 2099));
> ```
</details>
Done.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 195 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyTT4_lg9bnGd6OLQ0:-KMz0NpC7B6PDY9ycTh0:-499146009) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L195)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You could store the result of `temp_date + temp_time` into a const time_t variable to get rid of repetition here.
</details>
Done.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 202 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyTKbtdPUOe5NYT8Q8:-KMz0KQHg69Tmbti1tn3:2022736145) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L202)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> This should explain why they suck.
</details>
Removed function.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp, line 204 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyTC_29b3d52kCg0zc:-KMz0QKrcyGC7RU7Ns6K:2022736145) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.cpp#L204)):*
<details><summary><i>Previously, lioncash (Mat M.) wrote…</i></summary>
> You should probably not use names that differ this closely to one another. The parameter itself should likely be a time_t.
</details>
Removed function.

---

*[Source/Core/DolphinWX/Config/AdvancedConfigPane.h, line 38 \[r7\]](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007#-KMyP7kwM4YDWF5hHr2V:-KMz0RjC9Fk2TsBAHrmF:-1150875102) ([raw file](https://github.com/dolphin-emu/dolphin/blob/c581d3a8f9c0944d16d70029461c87328d24ab4d/Source/Core/DolphinWX/Config/AdvancedConfigPane.h#L38)):*
<details><summary><i>Previously, JosJuice wrote…</i></summary>
> These two need an `m_` prefix.
> 
> Actually, why do these exist at all? You can call m_custom_rtc_date_picker->GetValue().GetTicks()` in the places that use `temp_date` right now, and respectively for time. Keeping around extra state just makes the code harder to reason about.
</details>
Added the m_ prefix.

---


*Comments from [Reviewable](https://reviewable.io:443/reviews/dolphin-emu/dolphin/4007)*
<!-- Sent from Reviewable.io -->

Fallcrest commented:
@RisingFog When you set a custom time, does it stay locked in that time or does the emulated system clock move forwards like normal, just starting at the custom time?

JMC47 commented:
It moves forward from my testing.

RisingFog commented:
@Fallcrest it moves forward

Fallcrest commented:
Could you have it so that you can see the current system time in the settings window during gameplay? Since the game can't be interacted with when the settings window is open, I guess there would have to be some sort of pop-up with the time on it.

Would it be potentially useful for (even just for testing purposes) to be able to lock the time?

RisingFog commented:
@Fallcrest I can add an OSD toggle for system time perhaps.

I'm not sure if locking the time would be a good idea. GC/Wii always has running time.

Fallcrest commented:
@RisingFog An overlay would be a good solution, thanks.

JosJuice commented:
@RisingFog If you want the time to always be running forward, why does this PR let you change the time to anything at all while emulation is running?

RisingFog commented:
@JosJuice this is a good point.

RisingFog commented:
@Fallcrest I've added the OSD for system clock in #4036

Fallcrest commented:
@RisingFog I saw that, thanks. It should be valuable for those who need to know those values.
